packages/api/tests/lib/test-granules.js

```diff
@@ -91,3 +282,221 @@ test('getGranuleProductVolume() returns correct product volume', (t) => {
+  );
+});
+
+test.serial('deleteGranuleAndFiles() removes files from PG and S3', async (t) => {
+  const { newPgGranule, newDynamoGranule, files } = await createGranuleAndFiles(
+    t.context.knex,
+    t.context.testPgCollection.cumulus_id,
+    false
+  );
+
+  await deleteGranuleAndFiles({
+    knex: t.context.knex,
+    dynamoGranule: newDynamoGranule,
+    pgGranule: newPgGranule,
+    granuleModelClient: granuleModel,
+  });
+
+  // verify the files are deleted from S3. No need to check the Postgres files.
+  // If the granule was successfully deleted, the postgres
+  // files will have been as well. Files have a fk which would
+  // prevent the granule from being deleted if files referencing it
+  // still exist.
+  /* eslint-disable no-await-in-loop */
+  for (let i = 0; i < files.length; i += 1) {
+    const file = files[i];
+    t.false(await s3ObjectExists({ Bucket: file.bucket, Key: file.key }));
+  }
+  /* eslint-enable no-await-in-loop */
+});
+
+test.serial('deleteGranuleAndFiles() succeeds if a file is not present in S3', async (t) => {
+  const granuleId = randomId('granule');
+  const files = [
+    {
+      bucket: process.env.system_bucket,
+      fileName: `${granuleId}.hdf`,
+      key: randomString(),
+    },
+  ];
+
+  // Create Dynamo granule
+  const newGranule = fakeGranuleFactoryV2({ granuleId: granuleId, status: 'failed' });
+  newGranule.published = false;
+  newGranule.files = files;
+  await granuleModel.create(newGranule);
+
+  // create PG granule
+  const fakePGGranule = fakeGranuleRecordFactory(
+    {
+      granule_id: granuleId,
+      status: 'failed',
+      collection_cumulus_id: t.context.testPgCollection.cumulus_id,
+    }
+  );
+  fakePGGranule.published = false;
+  const [granuleCumulusId] = await granulePgModel.create(t.context.knex, fakePGGranule);
+
+  const newPgGranule = await granulePgModel.get(t.context.knex, { cumulus_id: granuleCumulusId });
+  const newDynamoGranule = await granuleModel.get({ granuleId: newGranule.granuleId });
+
+  await deleteGranuleAndFiles({
+    knex: t.context.knex,
+    dynamoGranule: newDynamoGranule,
+    pgGranule: newPgGranule,
+    granuleModelClient: granuleModel,
+  });
+
+  // Check Dynamo and RDS. The granule should have been removed from both.
+  await t.throwsAsync(
+    granuleModel.get({ granuleId: newDynamoGranule.granuleId }),
+    { instanceOf: RecordDoesNotExist }
+  );
+
+  await t.throwsAsync(
+    granulePgModel.get(t.context.knex, { granule_id: newPgGranule.granule_id }),
+    { instanceOf: RecordDoesNotExist }
+  );
+});
+
+test.serial('deleteGranuleAndFiles() will not delete a granule or its S3 files if the PG file delete fails', async (t) => {
+  const { newPgGranule, newDynamoGranule, files } = await createGranuleAndFiles(
+    t.context.knex,
+    t.context.testPgCollection.cumulus_id,
+    false
+  );
+
+  const mockFileModel = {
+    tableName: 'files',
+    delete: () => {
+      throw new Error('Delete failed');
+    },
+  };
```

Comment:
@botanical these tests are based on your comments around "failed" deletes. My thinking is that we want to handle actual errors that are thrown from the model or knex. To your point, it doesn't handle cases where "0" is returned from the delete() but I think that might be right? If delete() returns 0, it was technically a successful transaction but no rows were affected.

I also think that's probably an unlikely case where we actually return 0 (though feel free to push on that). The record we're passing to `delete()` should have been just fetched from the database in the previous step so I think its unlikely that it won't be found in delete? Definitely double-check my assumptions here.

Reply:
Yeah, I thought that "0" indicated a failure but, as you point out, it just means none of the rows were affected but it is still a successful transaction. I think your assumption is correct as well since we do fetch the record from the database before we attempt to delete. 

Thanks for considering this more and broadening my understanding on this!

Reply:
Likewise! It's an interesting question. And I'm still open to pushback if you or anyone else finds later that this isn't what we're looking for. I'll keep an eye out myself.

Reply:
re: my comments above re: the delete logic - I think we want to *for now* assume that some legit deletes might be called where there wasn't actually a record during migration. 

Once we go phase 2, a 0 result could be interpreted as a failure to find if our usage of it is consistent across endpoints/bulk delete, etc.